himalaya/internal/repair/repair.go

```diff
@@ -0,0 +1,365 @@
+
+	newManifest, err := manifest.NewDefaultManifest(r.ls, false)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	err = newManifest.Add(ctx, manifest.RootPath, manifest.NewEntry(
+		swarm.ZeroAddress,
+		map[string]string{
+			manifest.WebsiteIndexDocumentSuffixKey: oldEntry.mtdt.Filename,
+		},
+	))
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	err = newManifest.Add(
+		ctx,
+		oldEntry.mtdt.Filename,
+		manifest.NewEntry(oldEntry.e.Reference(), map[string]string{
+			manifest.EntryMetadataFilenameKey:    oldEntry.mtdt.Filename,
+			manifest.EntryMetadataContentTypeKey: oldEntry.mtdt.MimeType,
+		}),
+	)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	newReference, err := newManifest.Store(ctx)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	r.logger.Debugf("Created new file manifest with reference %s", newReference.String())
+
+	return newReference, nil
+}
+
+// DirectoryRepair takes in an older directory reference and creates a new manifest which contains
+// all the files and the metadata. This reference can be then used to query the /bzz endpoint to
+// serve the index document or /bzz/{reference}/{path} to query individual files
+//
+// Old Entry:
+// mantaray manifest -> Root Node (/) -> Metadata (index file/error file)
+//                   |
+//                   |-> file entry -> collection -> file reference -> file bytes
+//                                               |
+//                                               |-> metadata reference -> metadata bytes
+// New Entry:
+// mantaray manifest -> Root Node (/) -> Metadata (index file)
+//                  |
+//                  |-> file entry -> Metadata (Filename, ContentType)
+//                                |
+//                                |-> File reference
+//
+func DirectoryRepair(ctx context.Context, addr swarm.Address, opts ...Option) (swarm.Address, error) {
+	r := newWithOptions(opts...)
+
+	dir, err := r.getOldDirectoryEntry(ctx, addr)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+loop:
+	for {
+		select {
+		case f, ok := <-dir.filesC:
+			if !ok {
+				break loop
+			}
+			r.updater.Update(fmt.Sprintf("Updating reference for file %s", f.mtdt.Filename))
+			err := dir.m.Add(
+				ctx,
+				f.filepath,
+				manifest.NewEntry(f.e.Reference(), map[string]string{
+					manifest.EntryMetadataFilenameKey:    f.mtdt.Filename,
+					manifest.EntryMetadataContentTypeKey: f.mtdt.MimeType,
+				}),
+			)
+			if err != nil {
+				return swarm.ZeroAddress, err
+			}
+		case e, ok := <-dir.errC:
+			if !ok {
+				break loop
+			}
+			return swarm.ZeroAddress, e
+		case <-ctx.Done():
+			if ctx.Err() != nil {
+				return swarm.ZeroAddress, ctx.Err()
+			}
+			break loop
+		}
+	}
+
+	newReference, err := dir.m.Store(ctx)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	r.logger.Debugf("Created new directory manifest with reference %s", newReference.String())
```

Comment:
use Update?

Reply:
Again, I avoided this to prevent redundant messages on UI